cli/src/diff_util.rs

```diff
@@ -2067,60 +2078,111 @@ pub fn show_diff_stats(
-            let len = (stat.removed as f64 * factor).ceil() as usize;
-            (bar_length - len, len)
+            "(binary)".len()
         };
+        max_path_width = max_path_width.min(available_width.saturating_sub(width));
+    }
+
+    // Now that we've chosen the path width, use the rest of the space for the ++--
+    // bar.
+    let max_bar_width =
+        available_width.saturating_sub(max_path_width + diff_number_width + " ".len());
+    let factor = match max_diffs {
+        Some(max) if max > max_bar_width => max_bar_width as f64 / max as f64,
+        _ => 1.0,
+    };
+
+    for (stat, ui_path) in iter::zip(stats.entries(), &ui_paths) {
         // replace start of path with ellipsis if the path is too long
         let (path, path_width) = text_util::elide_start(ui_path, "...", max_path_width);
         let path_pad_width = max_path_width - path_width;
         write!(
             formatter,
-            "{path}{:path_pad_width$} | {:>number_width$}{}",
+            "{path}{:path_pad_width$} | ",
             "", // pad to max_path_width
-            stat.added + stat.removed,
-            if bar_added + bar_removed > 0 { " " } else { "" },
         )?;
-        write!(formatter.labeled("added"), "{}", "+".repeat(bar_added))?;
-        writeln!(formatter.labeled("removed"), "{}", "-".repeat(bar_removed))?;
+        if let Some((added, removed)) = stat.added_removed {
+            let bar_length = ((added + removed) as f64 * factor) as usize;
+            // If neither adds nor removes are present, bar length should be zero.
+            // If only one is present, bar length should be at least 1.
+            // If both are present, bar length should be at least 2.
+            //
+            // Fractional space after scaling is given to whichever of adds/removes is
+            // smaller, to show at least one tick for small (but nonzero) counts.
+            let bar_length = bar_length.max((added > 0) as usize + (removed > 0) as usize);
+            let (bar_added, bar_removed) = if added < removed {
+                let len = (added as f64 * factor).ceil() as usize;
+                (len, bar_length - len)
+            } else {
+                let len = (removed as f64 * factor).ceil() as usize;
+                (bar_length - len, len)
+            };
+            write!(
+                formatter,
+                "{:>diff_number_width$}{}",
+                added + removed,
+                if bar_added + bar_removed > 0 { " " } else { "" },
+            )?;
+            write!(formatter.labeled("added"), "{}", "+".repeat(bar_added))?;
+            writeln!(formatter.labeled("removed"), "{}", "-".repeat(bar_removed))?;
+        } else {
+            write!(formatter.labeled("binary"), "(binary)")?;
+            if stat.bytes_delta != 0 {
+                if stat.bytes_delta < 0 {
+                    write!(formatter.labeled("removed"), " {:+}", stat.bytes_delta)?;
+                } else if stat.bytes_delta > 0 {
```

Comment:
nit: just `} else {` (because we'll print " bytes" unconditionally)?

It might be also good to do `let label = if ...` and `write!(formatter.labeled(label), ..)`.